pkg/agent/nplagent/k8s/annotations.go

```diff
@@ -0,0 +1,165 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/vmware-tanzu/antrea/pkg/util/env"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+)
+
+const NPLAnnotationStr = "npl.antrea.io"
+
+type NPLEPAnnotation struct {
+	PodPort  string `json:"podPort"`
+	NodeIP   string `json:"nodeIP"`
+	NodePort string `json:"nodePort"`
+}
+
+func stringify(serialize interface{}) string {
+	jsonMarshalled, _ := json.Marshal(serialize)
+	return string(jsonMarshalled)
+}
+
+func isNodePortInAnnotation(s []NPLEPAnnotation, nodeport string) bool {
+	for _, i := range s {
+		if i.NodePort == nodeport {
+			return true
+		}
+	}
+	return false
+}
+
+func assignPodAnnotation(pod *corev1.Pod, containerPort, nodeIP, nodePort string) {
+	var err error
+	current := pod.Annotations
+	if current == nil {
+		current = make(map[string]string)
+	}
+
+	klog.V(0).Infof("Building annotation for pod: %s\tport: %s --> %s:%s", pod.Name, containerPort, nodeIP, nodePort)
+
+	var annotations []NPLEPAnnotation
+	// nplEP annotation exists
+	if current[NPLAnnotationStr] != "" {
+		if err = json.Unmarshal([]byte(current[NPLAnnotationStr]), &annotations); err != nil {
+			klog.Warningf("Unable to unmarshal NPLEP annotation: %v", current[NPLAnnotationStr])
+		}
+
+		if !isNodePortInAnnotation(annotations, nodePort) {
+			annotations = append(annotations, NPLEPAnnotation{
+				PodPort:  containerPort,
+				NodeIP:   nodeIP,
+				NodePort: nodePort,
+			})
+		}
+	} else {
+		annotations = []NPLEPAnnotation{{
+			PodPort:  containerPort,
+			NodeIP:   nodeIP,
+			NodePort: nodePort,
+		}}
+	}
+
+	current[NPLAnnotationStr] = stringify(annotations)
+	pod.Annotations = current
+}
+
+func removeFromPodAnnotation(pod *corev1.Pod, containerPort string) {
+	var err error
+	current := pod.Annotations
+
+	klog.V(0).Infof("Removing annotation from pod: %s\tport: %s", pod.Name, containerPort)
+	var annotations []NPLEPAnnotation
+	if err = json.Unmarshal([]byte(current[NPLAnnotationStr]), &annotations); err != nil {
+		klog.Warningf("Unable to unmarshal NPLEP annotation: %v", current[NPLAnnotationStr])
+		return
+	}
+
+	for i, ann := range annotations {
+		if ann.PodPort == containerPort {
+			annotations = append(annotations[:i], annotations[i+1:]...)
+			break
+		}
+	}
+
+	current[NPLAnnotationStr] = stringify(annotations)
+	pod.Annotations = current
+}
+
+// RemoveNPLAnnotationFromPods Removes npl annotations from all pods
```

Comment:
Removes -> removes

pods -> Pods
Probably add '.' at EOL.
Could you check other comments and logs?